### DISCUSSION TITLE
Add a jsonschema/pydantic based forms element

### DISCUSSION THREAD

**Discussion by eli-kha** [zauberzeug/nicegui#378] [Ideas / Feature Requests]
Hi,

I have a project that I'm working on that manages various devices and we use jsonschema and pydantic to represent configurations.
This allows use uniform APIs based on JSON and aligns well with FastAPI's use of pydantic.

I think that an element that generates an input form based on a pydantic model or pure jsonschema can be very useful.
From my research I found that both Django and FastAPI have extensions for this but both are not maintained:

- https://reforms.boardpack.org/
- https://github.com/Aristotle-Metadata-Enterprises/django-jsonforms

If you go directly to Javascript:

- [JSONForms ](https://jsonforms.io/) is the most maintained library that I found and it has declared Vue support.
- django-jsonforms uses [json-editor](https://github.com/json-editor/json-editor) which is also maintained and seems much leaner but I couldn't find anything explicit about Vue.

I tried looking into adding it myself based on the Vue examples and couldn't quite figure it out.

- json-editor seems much simpler and django-jsonforms offers decent reference code.
- JSONForms seems to offer better UI control without needing to change the HTML/JS code through UI_schema

If you give me enough pointers I can try implementing it

**Comment by rodja** [comment]
Nice! I really like the idea of generating input UI directly from a model. I had a brief look at the links you provided and would suggest an entirely other approach: Write it all in Python. Do not bother with Vue, JS, Jinja and everything else. Because you have NiceGUI. Just create a method `create_input_form(model)` which parses the model and creates `ui.input()`, `ui.number()`, ... elements for every field. Then you can call the method at the position in your UI where you want to show the form.

**Reply by eli-kha** [reply]
There are Pros and Cons to both approaches and it's kind of a mirror image.

### Pure python:
**Pro:**
- It would maintain the visual style of the interface.
- Alignment with Pydantic for the validation (you can just use their validators)

**Con:**
- Build from scratch and maintain the compatibility with the Model functionality over time (map types to UI elements, etc)
- Build and maintaint a DSL (like ui_schema) to allow designing the form
- Validation will be in the backend, which means it will be less user friendly (you won't have instant feedback near the form input field itself)

Going with a JS library will be a mirror image.
I'll see if I can find some time to prototype it.

**Comment by rodja** [comment]
Maybe this is a good small external package which builds upon standard NiceGUI?

**Reply by Dronakurl** [reply]
Here it is :-)

https://github.com/Dronakurl/nicecrud

I needed something similar for my project and figured, I can share it as an extra package.
Btw, I called it "niceguicrud" on pypi as "nicecrud" was "too similar" to idk what kind of package.  
If it is not ok with you guys that it now starts with "nicegui", please let me know!

**Reply by rodja** [reply]
Looks cool. The name is fine. Could you add it to https://github.com/zauberzeug/nicegui/wiki/?

**Comment by eli-kha** [comment]
I came around to the preference to use python. I helps maintain the theme and behavior.

Here is an initial implementation that looks ugly and is limited but works for my current use-case:
https://gist.github.com/eli-kha/9e13144bb8c8e7efd98cc9a47985fbe4

I'm already seeing a gap with the fact that I don't return the handlers to the input elements since I can't turn on the error state after a failed validation.

Feedback would be greatly appreciated.

**Comment by hasansezertasan** [comment]
It sounds like a great idea.

Pydantic models support generating JSON Schema out-of-box and there are some useful ORM/ODM packages like SQLModel and ODMantic that is directly built on top of Pydantic, which they probably support generating JSON Schema. To this end, generating a form for almost any kind of model is fairly easy.

Is this feature supported by NiceGUI or will it be?

**Reply by rodja** [reply]
Have you seen https://github.com/zauberzeug/nicegui/discussions/378#discussioncomment-9640692?

**Reply by hasansezertasan** [reply]
> Have you seen [#378 (reply in thread)](https://github.com/zauberzeug/nicegui/discussions/378#discussioncomment-9640692)?

Got it, thanks ☮️.

**Comment by simonryf** [comment]
Nice work @Dronakurl - I have not checked the entire codebase yet but I'm pretty sure I'm going to use it at some point. I have added it to [https://deepwiki.com/Dronakurl/nicecrud](https://deepwiki.com/Dronakurl/nicecrud) 😛

For those who can't or don't want to use the whole CRUD. I vibe-wrote a Class to validate single or multiple fields. I have not yet tested it thoroughly but so far it seems to work ;-). I've also incorporated the [https://github.com/zauberzeug/nicegui/discussions/1791?utm_source=chatgpt.com#discussioncomment-7270147](https://github.com/zauberzeug/nicegui/discussions/1791?utm_source=chatgpt.com#discussioncomment-7270147) so we can use it on a submit button.

```python
from typing import Any, Callable
from pydantic import BaseModel, ValidationError, TypeAdapter


class PydanticNiceValidator:
    """
    Create NiceGUI-compatible field validators from a Pydantic model and expose an overall validity flag.

    This adapter produces callables for NiceGUI's `validation` parameter. Each callable gets the field's
    current value and returns `None` (valid) or an error message string (invalid).

    It also maintains an `is_valid` attribute you can bind UI components to, such as enabling a submit button.
    `is_valid` is kept in sync after every validation call:

    - **With `ctx` (model instance or callable)**: the entire model is validated, so cross-field validators apply.
    - **Without `ctx`**: it aggregates the success/failure of the fields that this instance manages
      (i.e. for which you've called :meth:`validator_for`). Cross-field logic is *not* enforced in this mode.

    Hint:
        If your Pydantic model sets `model_config = ConfigDict(validate_assignment=True)`, assignments
        (e.g. via `bind_value`) might raise immediately. Decide whether you want that *in addition* to
        the UI-level feedback. If not, disable it as shown in the example.

    Example:
        Full NiceGUI usage with cross-field validation, re-validation of dependent fields, and button enablement:

        ```python
        from nicegui import ui
        from pydantic import BaseModel, Field, field_validator, ConfigDict

        class User(BaseModel):
            model_config = ConfigDict(validate_assignment=False)  # see docstring hint above
            username: str = Field(min_length=3, max_length=20, pattern=r'^[a-z0-9_]+$')
            password1: str = Field(min_length=8)
            password2: str

            @field_validator('password2')
            def passwords_match(cls, v, info):
                if 'password1' in info.data and v != info.data['password1']:
                    raise ValueError('Passwords must match')
                return v

        user = User(username='', password1='', password2='')

        validator = PydanticNiceValidator(User, ctx=user)

        username_in = ui.input('Username', validation=validator.validator_for('username')).bind_value(user, 'username')
        pw1_in = ui.input('Password', password=True, validation=validator.validator_for('password1')).bind_value(user, 'password1')
        pw2_in = ui.input('Repeat Password', password=True, validation=validator.validator_for('password2')).bind_value(user, 'password2')

        # Ensure pw2 is rechecked when pw1 changes:
        pw1_in.on('update:model-value', lambda e: pw2_in.validate())

        submit_btn = ui.button('Submit', on_click=lambda: print('Submitted:', user.model_dump()))
        submit_btn.bind_enabled_from(validator, 'is_valid')

        ui.run()
        ```

        Single-field usage *without* ctx (`is_valid` still works, but only for that field):

        ```python
        email_validator = PydanticNiceValidator(EmailModel)          # ctx=None
        email_in = ui.input('Email', validation=email_validator.validator_for('email'))

        submit = ui.button('Send', on_click=lambda: print('sending…'))
        submit.bind_enabled_from(email_validator, 'is_valid')
        ```

        For a standalone and simple field validator, you can use the static method `single`:

        ```python
        email_validator = PydanticNiceValidator.single(EmailModel, 'email')
        email_in = ui.input('Email', validation=email_validator)
        ```

    Attributes:
        model_cls: The Pydantic model class used.
        is_valid: `True` if valid, `False` if any managed field is invalid, `None` before any validation runs.

    Args:
        model_cls: Pydantic `BaseModel` subclass to validate against.
        ctx: Optionally one of:
            - a `BaseModel` instance to validate the full model each time,
            - a callable returning `dict[str, Any]` with the current values,
            - or `None` to validate fields independently.

    Raises:
        ValueError: If `ctx` is neither `None`, a `BaseModel` instance, nor a callable.
    """

    def __init__(
        self,
        model_cls: type[BaseModel],
        ctx: BaseModel | Callable[[], dict[str, Any]] | None = None,
    ) -> None:
        self.model_cls = model_cls
        self.is_valid: bool | None = None  # starts unknown until first validation
        self._field_valid: dict[str, bool] = {}  # track per-field validity (ctx=None mode)

        if isinstance(ctx, BaseModel):
            self._ctx_model = ctx
            self._ctx_fn: Callable[[], dict[str, Any]] | None = lambda: self._ctx_model.model_dump()
        elif callable(ctx):
            self._ctx_model = None
            self._ctx_fn = ctx
        elif ctx is None:
            self._ctx_model = None
            self._ctx_fn = None
        else:
            raise ValueError("ctx must be a BaseModel instance, a callable, or None")

        self._adapters: dict[str, TypeAdapter[Any]] = {
            name: TypeAdapter(field.annotation) for name, field in model_cls.model_fields.items()
        }

    def validator_for(self, field_name: str) -> Callable[[Any], str | None]:
        """Return a NiceGUI validation callable for a single field.

        Args:
            field_name: Field name on `model_cls`.

        Returns:
            Callable[[Any], str | None]: pass to `ui.input(..., validation=...)`.

        Raises:
            KeyError: If `field_name` is not defined on the model.
        """
        if field_name not in self._adapters:
            raise KeyError(f"Unknown field '{field_name}' on model {self.model_cls.__name__!r}")

        adapter = self._adapters[field_name]
        # Initialize tracking; assume invalid until first successful check
        if field_name not in self._field_valid:
            self._field_valid[field_name] = False

        def _rule(value: Any) -> str | None:
            try:
                if self._ctx_fn:
                    data = self._ctx_fn().copy()
                    data[field_name] = value
                    self.model_cls.model_validate(data)
                    self._field_valid[field_name] = True
                else:
                    adapter.validate_python(value)
                    self._field_valid[field_name] = True

                self._recalc_is_valid()
                return None
            except ValidationError as e:
                self._field_valid[field_name] = False
                self._recalc_is_valid()
                for err in e.errors():
                    loc = err.get("loc", ())
                    if loc and loc[0] == field_name:
                        return err.get("msg", "Invalid value")

        return _rule

    def _recalc_is_valid(self) -> None:
        """Refresh the public is_valid flag."""
        if self._ctx_fn:
            try:
                self.model_cls.model_validate(self._ctx_fn())
                self.is_valid = True
            except ValidationError:
                self.is_valid = False
        else:
            # Aggregate only the fields this instance manages
            if not self._field_valid:
                self.is_valid = None
            else:
                self.is_valid = all(self._field_valid.values())

    @classmethod
    def single(cls, model_cls: type[BaseModel], field_name: str) -> Callable[[Any], str | None]:
        """Shortcut for a standalone field validator (no instance, no `is_valid`, no cross-field checks)."""
        field = model_cls.model_fields[field_name]
        adapter = TypeAdapter(field.annotation)

        def rule(value: Any) -> str | None:
            try:
                adapter.validate_python(value)
                return None
            except ValidationError as e:
                return e.errors()[0].get("msg", "Invalid value") if e.errors() else "Invalid value"

        return rule
```

**Comment by BlankAdventure** [comment]
I came here to mention nicecrud (https://github.com/Dronakurl/nicecrud) as well.

I've played around with it and its a great idea. I ended up not using it for a particular project though due to lack of flexibility in customizing the behviours of certain fields, and the overall visual quality of the layout -- but these things could all be worked I'm sure if people were interested in doing so :)
